Cryptography API docs (#3319)
dapr/docs: daprdocs/content/en/developing-applications/building-blocks/cryptography/howto-cryptography.md

---
type: docs
title: "How to: Use the cryptography APIs"
linkTitle: "How to: Use cryptography"
weight: 2000
description: "Learn how to encrypt and decrypt files"
---

> Now that you've read about [Cryptography as a Dapr building block]({{< ref cryptography-overview.md >}}), let's walk through using the [high-level cryptography APIs]({{< ref cryptography_api.md >}}) with the Dapr SDKs. 

Reviewer: ```suggestion
Now that you've read about [Cryptography as a Dapr building block]({{< ref cryptography-overview.md >}}), let's walk through using the [cryptography APIs]({{< ref cryptography_api.md >}}) with the SDKs. 
```

{{% alert title="Note" color="primary" %}}
  Dapr cryptography is currently in alpha.

{{% /alert %}}

> 

Reviewer 2: Here maybe we can add some notes:

**First:**

Explain the difference between high-level and subtle crypto APIs. The subtle crypto APIs are not available yet, and will be offered in 1.12.

- High-level APIs allow encrypting and decrypting data using the [Dapr Crypto Scheme v1](https://dapr.io/enc/v1) (@msfussell this link depends on dapr/website#76 being merged). This is an opinionated encryption scheme that is designed to use modern, safe cryptographic standards, and processes data (even large files) efficiently as a stream.
- Subtle APIs (will be available in 1.12) are low-level APIs that allow performing a variety of cryptographic operations (encryption and decryption, key wrapping and unwrapping, creating and verifying digital signatures) within the vaults (or the built-in cryptographic engine). These APIs can be used to implement custom solutions and/or compatibility with other schemes. However, they provide less "guardrails" so users need to be aware of what they're doing, including how to use the cryptographic operations safely (e.g. what parameters/options to use)

> Dapr Cryptography is currently in alpha.

Reviewer: ```suggestion
Dapr cryptography is currently in alpha.
```

## Encrypt

> To read and encrypt a file, add the `Encrypt` gRPC API to your project.

Reviewer 2: ```suggestion
Using the Dapr gRPC APIs in your project, you can encrypt a stream of data, such as a file.
```

{{< tabs "Go" >}}

{{% codetab %}}

<!--go-->

```go
out, err := sdkClient.Encrypt(context.Background(), rf, dapr.EncryptOptions{
	// These are the 3 required parameters
	ComponentName: "mycryptocomponent",
	KeyName:        "mykey",

> 	Algorithm:     "RSA",

Reviewer 2: ```suggestion
	// Name of the Dapr component (required)
	ComponentName: "mycryptocomponent",
	// Name of the key stored in the component (required)
	KeyName:        "mykey",
	// Algorithm used for wrapping the key, which must be supported by the key named above.
	// Options include: "RSA", "AES"
	Algorithm:     "RSA",
```

})
```

{{% /codetab %}}

{{< /tabs >}}

The following example puts the `Encrypt` API in context, with code that reads the file, encrypts it, then stores the result in another file.

{{< tabs "Go" >}}

{{% codetab %}}

<!--go-->

> ```go

Reviewer 2: I think it would be useful to also show an example of encrypting a string too, and storing the result in-memory. This is so users don't think this is _only_ for files.

```go
rf := strings.NewReader("Amor, ch’a nullo amato amar perdona, mi prese del costui piacer sì forte, che, come vedi, ancor non m’abbandona")

// Encrypt the data using Dapr
enc, err := sdkClient.Encrypt(context.Background(), rf, dapr.EncryptOptions{
	ComponentName: "mycryptocomponent",
	KeyName:        "mykey",
	Algorithm:     "RSA",
})
if err != nil {
	panic(err)
}

// Read the encrypted data into a byte slice
enc, err := io.ReadAll(enc)
if err != nil {
	panic(err)
}
```

// Input file, clear-text
rf, err := os.Open("input")
if err != nil {
	panic(err)
}
defer rf.Close()

// Output file, encrypted
wf, err := os.Create("output.enc")
if err != nil {
	panic(err)
}
defer wf.Close()

// Encrypt the data using Dapr
out, err := sdkClient.Encrypt(context.Background(), rf, dapr.EncryptOptions{
	// These are the 3 required parameters
	ComponentName: "mycryptocomponent",
	KeyName:        "mykey",
	Algorithm:     "RSA",
})
if err != nil {
	panic(err)
}

// Read the stream and copy it to the out file
n, err := io.Copy(wf, out)
if err != nil {
	panic(err)
}
fmt.Println("Written", n, "bytes")
```

{{% /codetab %}}

{{< /tabs >}}

## Decrypt

To decrypt a file, add the `Decrypt` gRPC API to your project.

{{< tabs "Go" >}}

{{% codetab %}}

<!--go-->

```go

> out, err := sdkClient.Decrypt(context.Background(), rf, dapr.EncryptOptions{

Reviewer 2: Can you add a sentence explaining that `out` is a stream too, so it can be written to file or read in memory as in the examples above?

---
type: docs
title: "How to: Use the cryptography APIs"
linkTitle: "How to: Use cryptography"
weight: 2000
description: "Learn how to encrypt and decrypt files"
---

Now that you've read about [Cryptography as a Dapr building block]({{< ref cryptography-overview.md >}}), let's walk through using the cryptography APIs with the SDKs.  

{{% alert title="Note" color="primary" %}}
  Dapr cryptography is currently in alpha.

{{% /alert %}}

## Encrypt

Using the Dapr gRPC APIs in your project, you can encrypt a stream of data, such as a file.

{{< tabs "Go" >}}

{{% codetab %}}

<!--go-->

```go
out, err := sdkClient.Encrypt(context.Background(), rf, dapr.EncryptOptions{
	// Name of the Dapr component (required)
	ComponentName: "mycryptocomponent",
	// Name of the key stored in the component (required)

> 	KeyName:        "mykey",

Reviewer: ```suggestion
	KeyName:       "mykey",
```

	// Algorithm used for wrapping the key, which must be supported by the key named above.
	// Options include: "RSA", "AES"
	Algorithm:     "RSA",
})
```

{{% /codetab %}}

{{< /tabs >}}

The following example puts the `Encrypt` API in context, with code that reads the file, encrypts it, then stores the result in another file.

{{< tabs "Go" >}}

{{% codetab %}}

<!--go-->

```go
// Input file, clear-text
rf, err := os.Open("input")
if err != nil {
	panic(err)
}
defer rf.Close()

// Output file, encrypted
wf, err := os.Create("output.enc")
if err != nil {
	panic(err)
}
defer wf.Close()

// Encrypt the data using Dapr
out, err := sdkClient.Encrypt(context.Background(), rf, dapr.EncryptOptions{
	// These are the 3 required parameters
	ComponentName: "mycryptocomponent",

> 	KeyName:        "mykey",

Reviewer: ```suggestion
	KeyName:       "mykey",
```

	Algorithm:     "RSA",
})
if err != nil {
	panic(err)
}

// Read the stream and copy it to the out file
n, err := io.Copy(wf, out)
if err != nil {
	panic(err)
}
fmt.Println("Written", n, "bytes")
```

{{% /codetab %}}

{{< /tabs >}}

> The following example uses the `Ecrypt` API to encrypt a string.

Reviewer 2: ```suggestion
The following example uses the `Encrypt` API to encrypt a string.
```

{{< tabs "Go" >}}

{{% codetab %}}

<!--go-->

```go
// Input string
rf := strings.NewReader("Amor, ch’a nullo amato amar perdona, mi prese del costui piacer sì forte, che, come vedi, ancor non m’abbandona")

// Encrypt the data using Dapr
enc, err := sdkClient.Encrypt(context.Background(), rf, dapr.EncryptOptions{
	ComponentName: "mycryptocomponent",

> 	KeyName:        "mykey",

Reviewer: ```suggestion
	KeyName:       "mykey",
```
